ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py

```diff
@@ -0,0 +1,277 @@
+        return update_stats
+
+    def _update_batch(
+        self,
+        mini_batch_demo: Dict[str, np.ndarray],
+        mini_batch_policy: Dict[str, np.ndarray],
+    ) -> Dict[str, float]:
+        """
+        Helper method for update.
+        :param mini_batch_demo: A mini batch of expert trajectories
+        :param mini_batch_policy: A mini batch of trajectories sampled from the current policy
+        :return: Output from update process.
+        """
+        feed_dict = {
+            self.model.done_expert: mini_batch_demo["done"].reshape([-1, 1]),
+            self.model.done_policy: mini_batch_policy["done"].reshape([-1, 1]),
+            self.model.use_noise: [1.0],
+        }
+
+        if self.policy.use_continuous_act:
+            feed_dict[self.policy.model.selected_actions] = mini_batch_policy[
+                "actions"
+            ].reshape([-1, self.policy.model.act_size[0]])
+            feed_dict[self.model.action_in_expert] = mini_batch_demo["actions"].reshape(
+                [-1, self.policy.model.act_size[0]]
+            )
+        else:
+            feed_dict[self.policy.model.action_holder] = mini_batch_policy[
+                "actions"
+            ].reshape([-1, len(self.policy.model.act_size)])
+            feed_dict[self.model.action_in_expert] = mini_batch_demo["actions"].reshape(
+                [-1, len(self.policy.model.act_size)]
+            )
+
+        if self.policy.use_vis_obs > 0:
+            for i in range(len(self.policy.model.visual_in)):
+                policy_obs = mini_batch_policy["visual_obs%d" % i]
+                if self.policy.sequence_length > 1 and self.policy.use_recurrent:
+                    (_batch, _seq, _w, _h, _c) = policy_obs.shape
+                    feed_dict[self.policy.model.visual_in[i]] = policy_obs.reshape(
+                        [-1, _w, _h, _c]
+                    )
+                else:
+                    feed_dict[self.policy.model.visual_in[i]] = policy_obs
+
+                demo_obs = mini_batch_demo["visual_obs%d" % i]
+                if self.policy.sequence_length > 1 and self.policy.use_recurrent:
+                    (_batch, _seq, _w, _h, _c) = demo_obs.shape
+                    feed_dict[self.model.expert_visual_in[i]] = demo_obs.reshape(
+                        [-1, _w, _h, _c]
+                    )
+                else:
+                    feed_dict[self.model.expert_visual_in[i]] = demo_obs
+        if self.policy.use_vec_obs:
+            feed_dict[self.policy.model.vector_in] = mini_batch_policy[
+                "vector_obs"
+            ].reshape([-1, self.policy.vec_obs_size])
+            feed_dict[self.model.obs_in_expert] = mini_batch_demo["vector_obs"].reshape(
+                [-1, self.policy.vec_obs_size]
+            )
+
+        # for reporting
+        po, pe, zlss, zme, zmp = None, None, None, None, None
+        kl_loss = None
+        # end for reporting
+        if self.model.use_vail:
```

Comment:
Seems like this would be cleaner as something like
```python
out_dict = {
    "gail_loss": self.model.loss
    # ... same for update_batch, policy_estimate, expert_estimate
}

if self.model.use_vail:
    out_dict["kl"] = kl_loss
    # same for other VAIL variables
run_out = self.policy.sess.run(out_dict, feed_dict=feed_dict)

if if self.model.use_vail:
    self.update_beta(run_out["kl_loss"])

return run_out
```
Not sure if that changes the effect of `sess.run` though. And not sure if `out_dict` is the right name, but seems to be what we use elsewhere.

Reply:
I cleaned this up in the latest commit